tests/macros/src/main/scala/scala/macros/tests/sourcecode/SourceContext.scala

```diff
@@ -0,0 +1,77 @@
+package scala.macros.tests.sourcecode
+
+import scala.macros._
+import scala.language.implicitConversions
+
+// Unification of macro impl/defn forces us to put utilities in separate object
+private[sourcecode] object Util {
+  // TODO(olafur) replace with more typesafe typeOf[Line]
+  private def root: Term =
+    Term.Name("_root_").select("scala" :: "macros" :: "tests" :: "sourcecode" :: Nil)
+  def prefix(name: String): Term =
+    root.select(name :: Nil)
+
+  // Should this be handled by engines? I would claim that this is application specific
+  // for sourcecode.
+  def isSynthetic(s: Symbol)(implicit m: Mirror): Boolean = {
+    val name = s.name.value
+    name == "<init>" ||
+    name == "<root>" ||
+    (name.startsWith("<local ") && name.endsWith(">"))
+  }
+
+  def symbolName(s: Symbol)(implicit m: Mirror): String = s.name.value.stripSuffix("$")
+}
+
+case class File(value: String)
+object File {
+  implicit def generate: File = macro {
+    val path = enclosingPosition.input.path.toAbsolutePath.toString
+    Util.prefix("File").apply(Lit.String(path) :: Nil)
```

Comment:
For now, it seems just put the `path: String` in position make the usage easier.

Reply:
I don't think we should conflate `Input` and `Position`, they are two different concepts